boa/contracts/trace.py

```diff
@@ -0,0 +1,99 @@
+from dataclasses import dataclass
+from itertools import chain
+from pathlib import Path
+
+from boa.rpc import json
+from boa.util.abi import Address, abi_decode
+
+
+class TraceSource:
+    def format(self, input: bytes, output: bytes):
+        return f"{self}{self._format_input(input)}{self._format_output(output)}"
+
+    def _format_input(self, input: bytes):
+        decoded = abi_decode(self._input_schema, input)
+        args = [
+            f"{name} = {_to_str(d)}" for d, name in zip(decoded, self._argument_names)
+        ]
+        return f"({', '.join(args)})"
+
+    def _format_output(self, output: bytes):
+        if output == b"":
+            return " => None"
+        decoded = abi_decode(self._output_schema, output)
+        return f" => ({', '.join(_to_str(d) for d in decoded)})"
+
+    @property
+    def _input_schema(self) -> str:  # must be implemented by subclasses
+        raise NotImplementedError  # pragma: no cover
+
+    @property
+    def _argument_names(self) -> list[str]:  # must be implemented by subclasses
+        raise NotImplementedError  # pragma: no cover
+
+    @property
+    def _output_schema(self) -> str:  # must be implemented by subclasses
```

Comment:
ditto -- maybe something like `_return_schema`

Reply:
They don't need to be private I guess, it's pretty public info I think